mlflow/store/ftp_artifact_repo.py

```diff
@@ -26,52 +26,73 @@ def __init__(self, artifact_uri, client=None):
         if self.config['host'] is None:
             self.config['host'] = 'localhost'
 
-        if client:
-            self.ftp = client
-        else:
-            self.ftp = FTP()
-            self.ftp.connect(self.config['host'], self.config['port'])
-            self.ftp.login(self.config['username'], self.config['password'])
+        self.client = client
 
         super(FTPArtifactRepository, self).__init__(artifact_uri)
 
+    def get_ftp_client(self):
```

Comment:
I think if we import the contextmanager annotation (`from contextlib import contextmanager`) and add it to this method we can write it such that we use & close a fresh FTP client for each method. For example:

```
@contextmanager
def get_ftp_client(self):
  ftp = Ftp()
  ftp.connect(...)
  ftp.login(...)
  yield ftp
  ftp.close()
```

Then in each of the methods that uses an FTP client, we can write:
```
with self.get_ftp_client() as ftp:
  # Do things with ftp client. The client is automatically closed when the 'with' block
  # exits, including if we exit the 'with' block with an exception
  ftp.cwd(...)
```

The advantage of this is that we're guaranteed to use a fresh client in each method & close the client even if the method raises an exception. This will incur an extra overhead of logging into the FTP server in each method, but the overhead seems worthwhile if it allows us to avoid concerns about thread-safety.